glean-core/src/metrics/dual_labeled_counter.rs

```diff
@@ -0,0 +1,375 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::borrow::Cow;
+use std::char;
+use std::collections::HashMap;
+use std::io::Read;
+use std::sync::{Arc, Mutex};
+
+use crate::common_metric_data::{CommonMetricData, CommonMetricDataInternal, DynamicLabelType};
+use crate::error_recording::{record_error, test_get_num_recorded_errors, ErrorType};
+use crate::metrics::{CounterMetric, Metric, MetricType};
+use crate::Glean;
+
+const MAX_LABELS: usize = 16;
+const OTHER_LABEL: &str = "__other__";
+const MAX_LABEL_LENGTH: usize = 111;
+const RECORD_SEPARATOR: char = '\x1E';
+
+/// A dual labled metric, mirrors the Legacy keyed-categorical type.
+///
+/// Dual labled metrics allow recording multiple sub-metrics of the same type, in relation
+/// to two dimensions rather than the single label provided by the standard labeled type.
+#[derive(Debug)]
+pub struct DualLabeledCounterMetric {
+    keys: Option<Vec<Cow<'static, str>>>,
+    categories: Option<Vec<Cow<'static, str>>>,
```

Comment:
Do we need the `Option` here? Can we treat an empty Vec as no static keys/categories?
(Size-wise it doesn't matter, thanks to niche optimizations both types are 24 bytes)

Reply:
Ah I guess we do the same for single-labeled metrics. Fine keeping it consistent then.